### ISSUE TITLE
Add rank-normalized ESS and other variants

### ISSUE THREAD

**Issue by sethaxen** [TuringLang/MCMCDiagnosticTools.jl#22]
ESS is defined in the context of a specific estimate. For example, our current ESS implementations are all in the context of estimating the mean, so they will have problems whenever the mean/variance is not finite, e.g. for the Cauchy distribution. [Vehtari et al](https://arxiv.org/pdf/1903.08008.pdf) proposed several variants to the ESS/R-hat for diagnosing various problems that can manifest in posterior draws.

The different variants covered in the paper fall in the following categories

Splitting:
- split
- non-split

Pre-processing:
- rank-normalization to satisfy assumption of normality
- folding to avoid R-hat being fooled when chains have same location but different scales

Quantity being estimated:
- mean
- std
- quantile
- quantile interval
- mad

Some notes:
- splitting seems universally better than non-splitting, so I don't think we should support non-splitting.
- rank-normalization seems in general to be more robust than non-normalization, except for when one computes an MCSE; then one needs the un-rank-normalized version. So to me it makes sense for bulk-ESS (rank-normalized, unfolded ESS for estimate of mean) to be the default ESS method, which would also be consistent with ArviZ and Stan.
- Estimating ESSs for specific estimates generally requires pre- and post-processing of ESS estimates. Some examples given below (ignore the methods, this is not an API proposal):

```julia
ess_rhat(x, ::Val{:bulk}; kwargs...) = ess_rhat(rank_normalize(x); kwargs...)
function ess_rhat(x, ::Val{:tail}, p=0.90; kwargs...)
    return min(ess_rhat(x, Val(:quantile), (1-p)/2; kwargs...), ess_rhat(x, Val(:quantile), (1+p)/2; kwargs...))
end
ess_rhat(x, ::Val{:fold}; kwargs...) = ess_rhat(rank_normalize(fold(x)))
ess_rhat(x, ::Val{:quantile}, p; kwargs...) = ess_rhat(rank_normalize(qindicator(x, p)); kwargs...)
ess_rhat(x, ::Val{:median}; kwargs...) = ess_rhat(x, Val(:quantile), 0.5; kwargs...)
ess_rhat(x, ::Val{:mad}; kwargs...) = ess_rhat(rank_normalize(fold(x)), Val(:median); kwargs...)
function ess_rhat(x, ::Val{:qinterval}, pl, pu; kwargs...)
    return ess_rhat(rank_normalize(qindicatorrange(x, pl, pu)); kwargs...)
end
ess_rhat(x, ::Val{:mean}, p) = ess_rhat(x)
ess_rhat(x, ::Val{:std}, p) = min.(ess_rhat(x), ess_rhat(x.^2))
```

The current `AbstractESSMethod` approach doesn't give us the flexibility of specifying these variants. It also uses different types just more or less to specify differences in how the autocovariance is computed, whereas we can see there are more knobs a user might like to turn. Now would seem to be a good time to revisit this API.

**Comment by sethaxen** [comment]
Relates #10 and #4

**Comment by ParadaCarleton** [comment]
> splitting seems universally better than non-splitting, so I don't think we should support non-splitting.

I don't think that's true, so much as splitting measures something different from non-splitting. Splitting first makes `r-hat` a combined estimator of stationarity and mixing, while not splitting means `r-hat` only measures the latter. As an example, split r-hat will be worse at detecting multiple modes, because each chain will appear to have a stationary mean.

I think it's probably better as a default, although I've considered whether some other trend-based estimate of the stationarity would be better than splitting. I suspect directly trying to estimate an exponential trend and then using that to estimate the variance inflation would work better.

**Comment by ParadaCarleton** [comment]
I think that sounds great. Feel free to send me links to any relevant branches in Slack -- I was thinking of implementing something like this myself.

**Comment by sethaxen** [comment]
My (likely) final design is the following API. First there are the methods users are most likely to call, which should have the most complete documentation:
- `rhat(x)`: default R-hat; computes maximum of `rhat_bulk` and `rhat_tail`
- `ess(x)`: default bulk-ESS
- `ess(f, x)`:  ESS for estimator `f`
- `ess_rhat(x)`: default `ess(x)` and `rhat(x)` but more efficient

Then there are the less common methods, most of which are used by the above ones, which should be only lightly documented:
- `rhat(mean, x)`: mean (or basic) R-hat
- `rhat_bulk`: bulk-R-hat
- `rhat_tail`: tail-R-hat
- `ess_rhat_bulk(x)`: bulk-ESS and bulk-R-hat
- `ess_bulk(x)`: bulk-ESS. included solely for completeness.
- `ess_tail(x)`: tail-ESS

The idea is that calling `rhat`, `ess`, or `ess_rhat` without an estimator should give you the current recommended general-purpose diagnostic and clearly document what that is.

**Comment by ParadaCarleton** [comment]
Hmm, is there maybe a way to make this interface cleaner or more generalizable? e.g. `rhat_bulk(x)` could be replaced with `rhat(x; transform=rank_norm)`, where we provide a `rank_norm` transform for robustness. As a bonus, this extends to other kinds of transforms, e.g. if users want `rhat` for log-transformed data.

**Comment by sethaxen** [comment]
I don't see the benefit of a `transform` keyword, when we would just be applying the transform to `x` as a whole. They could just call `rhat(rank_norm(x))` or `rhat(log.(x))`, which they can do anyways. The main reason to have `rhat_bulk` and `rhat_tail` is that the user should not need to know what transform is used to construct each of those diagnostics (it's not useful for understanding the interpretation of the diagnostic), and requiring them to know or remember that poses a barrier to them computing the diagnostic.

Also `ess_tail` is constructed from the combination of two `ess` estimates so can't be computed with such a transformation.

Lastly, `rank_normalize` doesn't always convey robustness. e.g. we don't want the user combining `rank_normalize` with computing the ESS for `std`; they may compute an ESS then, but it won't be the ESS for `std`.

I do agree though that it's not ideal how many methods we have. e.g. for ArviZ and MCMCChains to completely extend our methods for their storage types, they would need to implement 10 methods (yikes!)

By comparison, posterior's API is:
- `ess_basic()`
- `ess_bulk()`
- `ess_tail()`
- `ess_quantile()`
- `ess_sd()`
- `mcse_mean()`
- `mcse_quantile()`
- `mcse_sd()`
- `rhat_basic()`
- `rhat()`

which is not too dissimilar to what I'm proposing here,

and Python ArviZ's API is:
- `rhat(x, method)`
- `ess(x, method)`

which is nice in its simplicity but requires each method be named. so e.g. `"mean"` and `"std"` are methods, as is `"bulk"`, instead of passing an estimator as we do in `mcse` and `ess_rhat` now.

**Comment by sethaxen** [comment]
I suppose we could adopt Python ArviZ's syntax for the base `rhat`, `ess`, and `ess_rhat` methods, with a `kind` or `method` keyword that we then use to dispatch to the other methods. This would be both easy for users to use and package developers to support.

**Comment by sethaxen** [comment]
Here's my best idea for a keyword API that works for `ess` and `rhat` but also is consistent with `mcse`:

- Instead of our current `mcse(estimator, x)` and `ess_rhat(estimator, x)` syntax, we would have `ess` and `mcse` accept an `estimator` keyword providing the estimator.
- For `ess`, only if `estimator` is not provided, a `type` keyword may be provided specifying `:bulk`, `:tail`, or `:basic`/`:mean`.
- For `rhat`, the `type` parameter accepts  `:bulk` `:tail`, `:basic`/`:mean` and `:rank` (the maximum of the two).
- `ess_rhat` would have the same `type` options as `rhat`, where `:rank` would return `:bulk` for ESS and `:rank` for R-hat.

Suggestions, @devmotion and @ParadaCarleton?

**Comment by sethaxen** [comment]
This week I will update the API similar to my previous comment, so we can hopefully make the breaking release.

**Comment by sethaxen** [comment]
Fixed by #72
